### DISCUSSION TITLE
Sobre Build Tools

### DISCUSSION THREAD

**Discussion by thiamsantos** [frontendbr/forum#323] [Perguntas e Respostas]
Eu acabei de ler esse [artigo](https://hackernoon.com/simple-build-tools-npm-scripts-vs-makefile-vs-runjs-31e578278162), e ele apresenta o [runjs](https://www.npmjs.com/package/runjs), que eu particularmente não conhecia, como uma excelente ferramenta de build que pode ser usada como alternativa ao gulp/grunt ou npm scripts.

Eu gostaria de saber de vocês:
- Vocês ainda utilizam gulp/grunt em seus projetos?
- Vale a pena usar npm scripts?
- Alguém já conhece ou já usou esse runjs?

**Comment by felipefialho** [comment]
Cara, continuo usando o Gulp, e sinceramente não vejo motivo nenhum pra não usar. É altamente configurável, com milhares de plugins úteis. 

Uso npm scripts para algumas tarefas apenas. 

Acredito que vou manter assim por um bom tempo ainda.

**Comment by YuriBrunetto** [comment]
Concordo com o @LFeh, uso o Gulp e nunca tive problemas com ele! O `npm scripts` uso quando preciso automatizar algo no projeto em si (acredito que todo mundo faça isso também).

**Comment by RuanAragao** [comment]
Também concordo com o @LFeh, estou muito bem até o momento com o Gulp, também acho que passarei um tempinho ainda com ele. Mas até o momento não testei o runjs.

**Comment by eduardojmatos** [comment]
@thiamsantos quase abro uma issue perguntando mais ou menos isso! 😃 

Hoje estamos fazendo um projeto do zero aqui no GetNinjas, e acabamos fazendo tudo via `npm scripts`. Mas estávamos discutindo exatamente isso agora à tarde, sobre usar ou não um task runner.

Eu particularmente tenho alguns pontos contra o uso dessas ferramentas, embora tenha usado várias, que são:

1- overlay desnecessário de tarefas básicas de bash script (por exemplo, vi vários módulos que são muito básicos, como `del`, que deleta um arquivo ou pasta). Com bash isso seria extremamente fácil usando `shelljs`;

2- dependência de uso do task runner way. Você precisa fazer exatamente como o task runner manda, da forma exata que ele funciona. Todos tem seu estilo próprio de declarar, empilhar e executar as tarefas. Se fugir um pouco disso começa a virar um código meio macarrônico (misturado com padrão definido e JavaScript com total liberdade);

3- dependência de plugins externos de forma exagerada. Pra qualquer coisa você precisa instalar um plugin do task runner que faz o wrapper da tarefa que você queria (ex: `gulp-csslint` que tem como dependência o `csslint`, que já tem uma interface de command line);

Entendo os pontos bons, por exemplo do Gulp, de fazer o uso de Stream pra executar as tarefas. É um fator bem importante, pois impacta a performance de execução. Mas olhando só nesse aspecto, ainda faz sentido usar ou não?

Deixo a dúvida e aguardo os comentários também 😄

**Comment by thiamsantos** [comment]
@eduardojmatos fico feliz de saber que não é só eu que estou em dúvida nesse aspecto :grin:

Eu acabei largando o gulp principalmente pela dependência dos plugins, daí se dava um problema eu não sabia se era o plugin do gulp, o pacote de que ele dependia ou se era o meu código que era o problema. Apesar disso, sinto um pouco falta da facilidade de escrever as tasks usando o gulp.

**Comment by fdaciuk** [comment]
> Apesar disso, sinto um pouco falta da facilidade de escrever as tasks usando o gulp.

Acho que esse é o ponto chave de continuar usando Gulp pra várias coisas. Aprendendo a usar streams, você consegue criar suas próprias tasks, sem depender de tantos plugins, ou ainda usar ele só pra aproveitar o watch de forma fácil pra rodar dependências de linha de comando com o `child_process/spawn` xD

Sobre _npm scripts_, eu gosto de usá-los como base para qualquer projeto: independente do task runner utilizado, eu centralizo todos os comandos no `package.json`, pois facilita pra qualquer um ver quais são as tasks disponíveis para aquele projeto. 

Isso também acaba deixando mais simples a portabilidade entre ferramentas, já que os comandos finais serão sempre npm scripts =)

**Comment by felipefialho** [comment]
É bem isso @fdaciuk: 

Gulp + npm scripts = 🍺

**Comment by eduardojmatos** [comment]
Acho que vou fazer um comparativo de performance de npm scripts versus Gulp versus webpack pra tirar essa duvida hehehe :)

**Comment by felipefialho** [comment]
Outro ponto, é que eu costumo usar o Gulp apenas para gerenciar as tasks de assets. Acho mais fácil organiza-las.

As tarefas de servidor, costumo usar npm scripts.

Dica: Use o [Concurrently](https://t.co/WyA97AdoAy) para rodar múltiplas tarefas em paralelo via npm scripts.  Tipo se precisar rodar o nodemon e o gulp, e não quiser instalar um plugin no gulp para o nodemon.

**Comment by fdaciuk** [comment]
Já usei o concurrently, mas não curto o stdout dele.. ele avacalha quando precisar ter saída colorida :P
Ae eu uso o `spawn` que mantém a saída original do programa =)

**Comment by baltazarparra** [comment]
Atualmente uso gulp, mas minhas tasks são bem básicas,
sass, concat e uglify. Estou pensando em compilar o sass nativamente, e passar o concat e o uglify do Js pro npm scripts, pois já tive dores de cabeça com versões de plugins em projetos mais antigos, e gostaria de evitar isso lá na frente. Faz sentido? @fdaciuk @eduardojmatos @LFeh 

Os problemas na real, vieram com o Grunt, que usávamos até ano passado, então os projetos mais antigos que precisam de manutenção, quase sempre dão problema.

**Comment by fdaciuk** [comment]
@baltazarparra vejo que faz sentido se isso, de fato, facilitar o seu workflow.
Essas ferramentas têm o propósito principal de ajudar a resolver certos problemas de foma fácil e centralizada.

Sobre as versões: o problema é que tem uma galera que não sabe trabalhar corretamente com _semantic versioning_, e acaba fazendo atualizações que quebram coisas, aplicando a flag de _patch_ :|

Mas acho que faz tudo sentido usar _npm scripts_ direto com as CLIs para essas tarefas. O grande problema é quando você precisa usar mais de uma com um único comando. Até hoje não encontrei uma ferramenta de `watch` de decente pra rodar duas ou mais tarefas ao mesmo tempo.

Por isso continuo usando o Gulp + `child_process/spawn` pra várias coisas =)

**Comment by bernardodiasc** [comment]
Uso https://sankhs.com/watchman.js/ + https://github.com/shelljs/shelljs numa tarefa de testes aqui, funciona bem.

**Comment by fdaciuk** [comment]
Legal @bernardodiasc! O `shelljs` acho que até faria sentido no meu caso, mas o `watchman` já seria uma outra interface que eu precisaria usar. Nesse caso, prefiro o `watch` do gulp que já é mais conhecido =)

**Comment by vitorbritto** [comment]
@thiamsantos tem um certo tempo que [escrevi este artigo](http://www.vitorbritto.com.br/blog/evolucao-automacao-de-tarefas/). De lá para cá, muita coisa aconteceu. Novas ferramentas surgiram e novos métodos também. =]

Você pode automatizar suas tarefas com:
- Shell Script
- makefile
- npm
- ShellJS
- MakeJS
- Grunt
- Gulp
- entre outras _tools_

### Sobre seus questionamentos:

**Vocês ainda utilizam gulp/grunt em seus projetos?**

Sim. No momento, utilizo mais o Gulp.

**Vale a pena usar npm scripts?**

IMHO, depende do projeto e do conhecimento da equipe que está lidando com o projeto.

**Alguém já conhece ou já usou esse runjs?**

Não conheço.

**Comment by cjpatoilo** [comment]
@thiamsantos Uso Gulp nos projetos da empresa e nos meus projetos ainda possuo alguns com Grunt. Ao poucos estou migrando os projetos com funcionalidades mais simples do Grunt para npm scripts. 

Sobre valer a pena usar npm scripts, bom, uso npm scripts há pouco tempo e já passei por problemas bem específicos ao tentar reproduzir tasks com o mesmo resultado que eu possuía no Grunt. Acho válida a experiência mas com certeza você deve encontrar um melhor suporte (dicas, documentação, tutorial..) no Grunt e no Gulp.

Sobre o runjs, primeira vez que ouço esse termo =) falando nisso, obrigado por compartilhar o artigo ;) 

@fdaciuk @LFeh Sobre a ferramenta para correr tasks em paralelo, vocês conhecem esse cara? https://www.npmjs.com/package/npm-run-all

@eduardojmatos Sobre o comparativo de performance, seria muito maneiro! +1

**Comment by fdaciuk** [comment]
@cjpatoilo já tinha ouvido falar, mas não cheguei a testar.. acho que é o momento, haha! Valeu a dica :dancer: :dancer: :sparkles:

**Comment by felipefialho** [comment]
@cjpatoilo Não conhecia! Valeu a dica.

**Comment by dsoldera** [comment]
Eu tb já tentei escrevet alguns scripts direto no Node, mas não tem a mesma facilidade e banco de exemplos na Comunidade como usar o Gulp ou o Grunt. As vezes pé só se vale a pena ficar reescrevendo a roda.
